File: languages/java/app/src/main/java/org/benchmarkplayground/benchmarkers/Benchmarker.java
package org.benchmarkplayground.benchmarkers;

import org.benchmarkplayground.utils.GeneralUtilities;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public abstract class Benchmarker {
    private static short MILLI_TO_NANO_FACTOR = 6;
    private static double NANO_PER_MILLI = Math.pow(10, MILLI_TO_NANO_FACTOR);

    protected String operationName;

    public abstract Long getOperationExecutionTime();

    public abstract void consumeInputFile(String inputFilePath) throws IOException;

    public String getOperationName() {
        return operationName;
    }

    public List<Long> getOperationExecutionResults(int executionCount) {
        List<Long> executionResults = new ArrayList<>(executionCount);
        for (int i = 0; i < executionCount; i++) {
            executionResults.add(getOperationExecutionTime());
        }
        return executionResults;
    }

    public void printBenchmarkAnalysis(int executionCount) {
        List<Long> executionResults = getOperationExecutionResults(executionCount);
        Long total = executionResults.stream().reduce(Long.valueOf(0), Long::sum);
        double formattedTime = GeneralUtilities.roundToDecimals(total / NANO_PER_MILLI, MILLI_TO_NANO_FACTOR);
        System.out.println(String.format("Java's %s execution time (over %d loops): %.6f ms", operationName, executionCount, formattedTime));
    }
}

File: languages/rust/src/benchmarkers/benchmarker.rs
/**
 * Contains base benchmarker trait that will be used by all operation benchmarkers.
 */
 
use std::{ops::Add, time::Duration};

use crate::utils::errors::BenchmarkerError;

pub trait Benchmarker {
    fn consume_input_file(&mut self, input_file_path: &str) -> Result<(), BenchmarkerError>;

    fn get_operation_name(&self) -> &str;

    fn get_operation_execution_time(&self) -> Duration;

    fn get_operation_execution_results(&self, execution_count: u32) -> Vec<Duration> {
        Vec::from_iter((0..execution_count).map(|_| self.get_operation_execution_time()))
    }

    fn print_benchmark_analysis(&self, execution_count: u32) -> Result<(), BenchmarkerError> {
        let execution_results = self.get_operation_execution_results(execution_count);
        let total_elapsed = execution_results.into_iter().fold(Duration::from_secs(0), |a, b| a.add(b));

        // Prints statistics
        let operation_name = self.get_operation_name();
        let milliseconds = total_elapsed.as_nanos() as f64 / 1_000_000 as f64;
        println!("Rust's {} execution time (over {:?} loops): {:.6} ms", operation_name, execution_count, milliseconds);

        Ok(())
    }
}
